File: Bai1.cpp
#include <iostream>
#include <vector>
#include<windows.h>

using namespace std;

class PriorityQueue {
private:
    vector<int> heap;  // Một mảng lưu trữ các phần tử trong cấu trúc đống

    // Hàm giúp điều chỉnh đống khi một phần tử được thêm vào
    void heapifyUp(int index) {
        while (index > 0) {
            int parentIndex = (index - 1) / 2;
            if (heap[index] > heap[parentIndex]) {  // So sánh giá trị của phần tử với cha
                swap(heap[index], heap[parentIndex]);  // Hoán đổi nếu cần
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    // Hàm giúp điều chỉnh đống khi một phần tử bị loại bỏ
    void heapifyDown(int index) {   
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;
        int largest = index;

        // Tìm ra phần tử lớn nhất giữa cha và hai con
        if (leftChild < heap.size() && heap[leftChild] > heap[largest]) {
            largest = leftChild;
        }
        if (rightChild < heap.size() && heap[rightChild] > heap[largest]) {
            largest = rightChild;
        }

        // Nếu cha không lớn nhất, hoán đổi và tiếp tục điều chỉnh đống
        if (largest != index) {
            swap(heap[index], heap[largest]);
            heapifyDown(largest);
        }
    }

public:
    // Thêm một phần tử vào hàng đợi ưu tiên
    void push(int value) {
        heap.push_back(value);  // Thêm vào cuối mảng
        heapifyUp(heap.size() - 1);  // Điều chỉnh để đảm bảo tính ưu tiên
    }

    // Trả về giá trị có độ ưu tiên cao nhất (đầu hàng đợi)
    int top() {
        if (!empty()) {
            return heap[0];
        }
        throw std::out_of_range("PriorityQueue is empty");
    }

    // Loại bỏ phần tử có độ ưu tiên cao nhất
    void pop() {
        if (!empty()) {
            heap[0] = heap.back();  // Thay thế phần tử đầu bằng phần tử cuối
            heap.pop_back();  // Loại bỏ phần tử cuối
            heapifyDown(0);  // Điều chỉnh để tái thiết lại đống
        }
    }

    // Kiểm tra xem hàng đợi ưu tiên có trống hay không
    bool empty() const {
        return heap.empty();
    }
};

int main() {

    PriorityQueue priorityQueue;

    // Chèn các giá trị vào hàng đợi
    priorityQueue.push(8);
    priorityQueue.push(3);
    priorityQueue.push(9);
    priorityQueue.push(1);
    priorityQueue.push(7);
    priorityQueue.push(5);
    priorityQueue.push(6);
    priorityQueue.push(4);

    //hiển thị dấu trong console
    SetConsoleOutputCP(65001);

    // Hiển thị các phần tử trong hàng đợi
    cout << "Các phần tử trong hàng đợi ưu tiên: ";
    while (!priorityQueue.empty()) {
        cout << priorityQueue.top() << " ";
        priorityQueue.pop();
    }
    
    cout << endl;

    return 0;
}

File: Bai1.py
# Bài 1. Cài đặt cấu trúc dữ liệu hàng đợi ưu tiên (sử dụng cấu trúc đống).
# Thử nghiệm chèn lần lượt vào hàng đợi các giá trị 8, 3, 9, 1, 7, 5, 6, 4 dùng thao tác Push();
# sau đó hiển thị các phần tử đó lên màn hình và rút tất cả các phần tử ra khỏi hàng đợi dùng thao tác pop().


class PriorityQueue:
    def __init__(self):
        self.queue = []                             # Khởi tạo một danh sách để lưu trữ các phần tử trong hàng đợi ưu tiên.

    def push(self, value):
        self.queue.append(value)                    # Thêm giá trị vào danh sách.
        self._heapify_up(len(self.queue) - 1)       # Gọi phương thức _heapify_up để cân bằng lại cây đống.

    def pop(self):
        if not self.is_empty():                     # Kiểm tra xem hàng đợi có rỗng hay không.
            if len(self.queue) == 1:
                return self.queue.pop()
            root = self.queue[0]                    # Lấy ra phần tử gốc của cây đống (phần tử nhỏ nhất).
            self.queue[0] = self.queue.pop()        # Gán phần tử cuối cùng của danh sách lên vị trí gốc.
            self._heapify_down(0)                   # Gọi phương thức _heapify_down để cân bằng lại cây đống.
            return root
        else:
            print("Hàng đợi ưu tiên đã rỗng.")

    def is_empty(self):
        return len(self.queue) == 0                 # Kiểm tra xem hàng đợi có rỗng hay không.

    def _heapify_up(self, index):
        while index > 0:
            parent_index = (index - 1) // 2         # Tìm vị trí của nút cha. (// python = % C++)
            if self.queue[index] < self.queue[parent_index]:
                self.queue[index], self.queue[parent_index] = self.queue[parent_index], self.queue[index]
                                                    # Hoán đổi phần tử con với phần tử cha nếu điều kiện thứ tự đống bị vi phạm.
                index = parent_index
            else:
                break

    def _heapify_down(self, index):
        while True:
            left_child_index = 2 * index + 1        # Tìm vị trí của nút con bên trái.
            right_child_index = 2 * index + 2       # Tìm vị trí của nút con bên phải.
            smallest = index                        # Gán vị trí của nút hiện tại là vị trí nhỏ nhất.

            if (left_child_index < len(self.queue) and
                self.queue[left_child_index] < self.queue[smallest]):
                smallest = left_child_index
            # So sánh với nút con bên trái và cập nhật vị trí nhỏ nhất nếu cần.

            if (right_child_index < len(self.queue) and
                self.queue[right_child_index] < self.queue[smallest]):
                smallest = right_child_index
            # So sánh với nút con bên phải và cập nhật vị trí nhỏ nhất nếu cần.

            if smallest != index:
                self.queue[index], self.queue[smallest] = self.queue[smallest], self.queue[index]
                # Hoán đổi phần tử hiện tại với phần tử nhỏ nhất nếu điều kiện thứ tự đống bị vi phạm.
                index = smallest
            else:
                break

priority_queue = PriorityQueue()

values_to_insert = [8, 3, 9, 1, 7, 5, 6, 4]
for value in values_to_insert:
    priority_queue.push(value)

print("Các phần tử trên hàng đợi ưu tiên:")
while not priority_queue.is_empty():
    item = priority_queue.pop()
    print(item,end=" ")